---
  html_document:
    toc: true
    highlight: zenburn
    theme: united
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=TRUE, highlight=TRUE, autodep=TRUE, warning=FALSE,
               error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='')
```

# Overview

```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(knitr)
project_summary = "../project-summary.csv"
counts_file = "../combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = read.table(project_summary, header=TRUE, sep=",")
rownames(summarydata) = summarydata$Name
summarydata = summarydata[order(summarydata$Name),]
counts = read.table(counts_file, header=TRUE, row.names="id")
counts = counts[, order(colnames(counts))]
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
    "rRNA.rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
    "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
    "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
    "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
    "complexity")
batch2 = read.table("../batch_2_samples.txt", sep="\t")
batch2$batch = 2
m = merge(summarydata, batch2, by.x="Name", by.y="V1", all.x=TRUE)
m$batch[is.na(m$batch)] = 1
summarydata = m
summarydata = summarydata[, ! colnames(summarydata) %in% c("species", "time")]
summarydata$tissue_status = as.factor(paste(summarydata$tissue, summarydata$status, sep="_"))
rownames(summarydata) = summarydata$Name

```

```{r heatmap-function}
get_heatmap_fn = function(summarydata) {
    # return the pheatmap function with or without metadata
    metadata = summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
    metadata = cbind(metadata, summarydata[, c("Mapping.Rate", "X.GC")])
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summarydata)
```

# Quality control metrics

## Summary statistics of summarydata
```{r summaries}

numeric_columns = c("X.GC", "Exonic.Rate", "rRNA.rate", "Intronic.Rate",
    "Intergenic.Rate", "Mapping.Rate", "Mapped", "Fragment.Length.Mean", "Genes.Detected",
    "Transcripts.Detected")
sumdf = data.frame(summary(summarydata[, numeric_columns]))


## Mapped reads
```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
```{r genes-detected-plot}
ggplot(summarydata, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Exonic mapping rate
```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA.rate)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Estimated fragment length of paired-end reads
```{r fragment-length-plot}
ggplot(summarydata, aes(x=Name, y=Fragment.Length.Mean)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("fragment length") + xlab("")
```

## Boxplot of log10 counts per gene
```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted$sample = reorder(melted$sample, colnames(counts))
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"), fontsize=6)
```

GC content seems to cluster with tissue type with the head and rest of body having lower
GC content than the spermatheca, MAG and atrium.

## Correlation (Spearman) heatmap of TMM-normalized counts
```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"), fontsize=6)
```

One sample, DT3_ALBI_Sp_M is very different from all of the other samples. This has the lower number of transcripts detected at 1.2k which is about 1/10th of the other samples. This sample likely has some kind of either contamination issue or PCR artifact happening, where a single gene or a small number of genes was sequenced repeatedly.



```{r outlier-sample, results='asis'}
kable(summarydata[summarydata$Name == "DT3_ALBI_Sp_M",], format="markdown")
```

as opposed to the first five samples sequenced:

```{r first-5, results='asis'}
kable(head(summarydata), format="markdown")
```

We should drop DT3_ALBI_Sp_M from the analysis.

```{r drop-DT3_ALBI_Sp_M}
summarydata = summarydata[!rownames(summarydata) == "DT3_ALBI_Sp_M",]
counts = counts[, !colnames(counts) == "DT3_ALBI_Sp_M"]
normalized_counts = normalized_counts[, !colnames(normalized_counts) == "DT3_ALBI_Sp_M"]
```

The overall exonic mapping rates is low:

```{r exonic-rate, results='asis'}
summary(summarydata$Exonic.Rate)
```

It isn't clear if these issues are due to the fact the genome/annotation for the mosquito
is poor or if there is a problem with the samples.

## MDS plot of TMM-normalized counts
```{r mds-normalized}
mds(normalized_counts, k=length(colnames(normalized_counts)) - 1)
```

There is pretty consistent separation of the head with M/V status, and maybe some separation
in MAG with M/V status, barring CT3, and an overall separation of tissue type. Interestingly,
atrium and rest of body samples look pretty similar to each other.

## Heatmap of top 30 most expressed genes
```{r top-count-genes, results='asis'}
select = order(rowMeans(counts),decreasing=TRUE)[1:30]
heatmap_fn(counts[select,])
```


# Differential expression
We'll look at differences that are driven by mating status in the same tissue. The way
we will do this is to create factors that describe both the tissue status and mating
status, and then do pairwise comparisons across those tissues. This will answer the
question of what genes are different in each specific tissue.

```{r de-setup}
library(DESeq2)
library(DEGreport)
library(vsn)
design = ~sex + tissue_status
condition = "tissue"
```

```{r deseq2-expression-analysis, results='asis'}
counts <- counts[rowSums(counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts, colData=summarydata, design = design)
dds = DESeq(dds)
```

Now that we have the differential expression calls using the tissue and status, we
can pull out the mating specific differences in each tissue.

### Atrium
```{r atrium-de}
atrium = results(dds, contrast=list(c("tissue_statusatrium_mated"),
                          c("tissue_statusatrium_virgin")))
write.table(atrium, file="atrium_deseq2.tsv", sep="\t", row.names=TRUE, col.names=TRUE,
            quote=FALSE)
atrium_de = subset(atrium, padj < 0.1)
```
There are `r nrow(atrium_de)` genes flagged as differentally expressed between
the atrium of mated and virgin samples, with `r nrow(subset(atrium_de, log2FoldChange > 0))`
upregulated in mated samples and `r nrow(subset(atrium_de, log2FoldChange < 0))` in atrium
samples.

There is quite a bit of variability between samples:

```{r atrium-ma}
DESeq2::plotMA(atrium)
```

```{r atrium-volcano}
atrium_stats = as.data.frame(atrium[, c(2,6)])
volcano_density_plot(atrium_stats, title=names(atrium_stats), lfc.cutoff=1.5)
```

### Head
```{r head-de}
head = results(dds, contrast=list(c("tissue_statushead_mated"),
                          c("tissue_statushead_virgin")))
write.table(head, file="head_deseq2.tsv", sep="\t", row.names=TRUE, col.names=TRUE,
            quote=FALSE)
head_de = subset(head, padj < 0.1)
```
There are `r nrow(head_de)` genes flagged as differentally expressed between
the atrium of mated and virgin samples, with `r nrow(subset(head_de, log2FoldChange > 0))`
upregulated in mated samples and `r nrow(subset(head_de, log2FoldChange < 0))` in atrium
samples.

There is quite a bit of variability between samples:

```{r head-ma}
DESeq2::plotMA(head)
```

```{r head-volcano}
head_stats = as.data.frame(head[, c(2,6)])
volcano_density_plot(head_stats, title=names(head_stats), lfc.cutoff=1.5)
```

### MAG
```{r MAG-de}
MAG = results(dds, contrast=list(c("tissue_statusMAG_mated"),
                          c("tissue_statusMAG_virgin")))
write.table(MAG, file="MAG_deseq2.tsv", sep="\t", row.names=TRUE, col.names=TRUE,
            quote=FALSE)
MAG_de = subset(MAG, padj < 0.1)
```
There are `r nrow(MAG_de)` genes flagged as differentally expressed between
the atrium of mated and virgin samples, with `r nrow(subset(MAG_de, log2FoldChange > 0))`
upregulated in mated samples and `r nrow(subset(MAG_de, log2FoldChange < 0))` in atrium
samples.

There is quite a bit of variability between samples:

```{r MAG-ma}
DESeq2::plotMA(MAG)
```

```{r MAG-volcano}
MAG_stats = as.data.frame(MAG[, c(2,6)])
volcano_density_plot(MAG_stats, title=names(MAG_stats), lfc.cutoff=1.5)
```

### body
```{r body-de}
summarydata$sex_tissue_status = paste(summarydata$sex, summarydata$tissue_status,
                                      sep="_")
design_body = ~sex_tissue_status
dds_body = DESeqDataSetFromMatrix(countData=counts, colData=summarydata,
                                  design = design_body)
dds_body = DESeq(dds_body)
female_body = results(dds_body, contrast=list(c("sex_tissue_statusfemale_body_mated"),
    c("sex_tissue_statusfemale_body_virgin")))
male_body = results(dds_body, contrast=list(c("sex_tissue_statusmale_body_mated"),
    c("sex_tissue_statusmale_body_virgin")))
write.table(male_body, file="male_body_deseq2.tsv", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
write.table(female_body, file="female_body_deseq2.tsv", sep="\t", row.names=TRUE, col.names=TRUE, quote=FALSE)
male_body_de = subset(male_body, padj < 0.1)
female_body_de = subset(female_body, padj < 0.1)
```
There are `r nrow(male_body_de)` genes flagged as differentally expressed between
the atrium of mated and virgin samples in the male, with
`r nrow(subset(male_body_de, log2FoldChange > 0))`
upregulated in mated samples and `r nrow(subset(male_body_de, log2FoldChange < 0))` in
the virgin samples samples.

There are `r nrow(female_body_de)` genes flagged as differentally expressed between
the atrium of mated and virgin samples in the female, with
`r nrow(subset(female_body_de, log2FoldChange > 0))`
upregulated in mated samples and `r nrow(subset(female_body_de, log2FoldChange < 0))` in
the virgin samples samples.

```{r body-ma}
DESeq2::plotMA(body)
```

```{r body-volcano}
body_stats = as.data.frame(body[, c(2,6)])
volcano_density_plot(body_stats, title=names(body_stats), lfc.cutoff=1.5)
```

### spermatheca
```{r spermatheca-de}
spermatheca = results(dds, contrast=list(c("tissue_statusspermatheca_mated"),
                          c("tissue_statusspermatheca_virgin")))
write.table(spermatheca, file="spermatheca_deseq2.tsv", sep="\t", row.names=TRUE, col.names=TRUE,
            quote=FALSE)
spermatheca_de = subset(spermatheca, padj < 0.1)
```
There are `r nrow(spermatheca_de)` genes flagged as differentally expressed between
the atrium of mated and virgin samples, with `r nrow(subset(spermatheca_de, log2FoldChange > 0))`
upregulated in mated samples and `r nrow(subset(spermatheca_de, log2FoldChange < 0))` in atrium
samples.

There is quite a bit of variability between samples:

```{r spermatheca-ma}
DESeq2::plotMA(spermatheca)
```

```{r spermatheca-volcano}
spermatheca_stats = as.data.frame(spermatheca[, c(2,6)])
volcano_density_plot(spermatheca_stats, title=names(spermatheca_stats), lfc.cutoff=1.5)
```

## PCA

Samples tend to cluster on the PCA plot by tissue and to a lesser degree mated/virgin status,
that is a good sign.

```{r pca}
rld = rlog(dds)
plotPCA(rld, intgroup=c("tissue", "status"))
```

## Effect of variance stabilization
There is a huge amount of dispersion between these samples.

```{r deseq-diagnostics, results='asis'}
par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

## Dispersion estimates

```{r dispersion-estimate}
plotDispEsts(dds)
```

## Tissue markers
One of the things we can do with this data is to produce a set of tissue-specific markers.
We're looking for a set a gene signatures that can be used to classify the different tissues.

We'll do NMF to pull out gene signatures for the different individual tissues.

```{r limma-counts}
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#000000")

library(edgeR)
library(limma)
library(NMF)
design = ~sex + tissue_status + sex:tissue_status
dge = DGEList(counts=counts)
dge = calcNormFactors(y)
voomed = voom(dge, plot=TRUE)
#tissue_metadata = subset(summarydata, tissue != "body")
tissue_metadata <-  summarydata
tissue_only = normalized_counts[, colnames(normalized_counts) %in% rownames(tissue_metadata)]
tissue_only = tissue_only[rowSums(tissue_only) > 1,]
adf_tissue = AnnotatedDataFrame(tissue_metadata[,c("sex", "tissue")],
    data.frame(labelDescription=c("sex", "tissue")))
x = ExpressionSet(as.matrix(tissue_only), adf_tissue)
nbasis <- length(unique(summarydata$tissue))
```

```{r run-nmf}
n = nmf(x, nbasis, nrun=100, .options=list(parallel=FALSE))
```

```{r nmf-plots}
consensus = consensusmap(n, annCol=x)
basismap(n, scale="r1", annColors=list(basis=cbPalette[4:8],
         consensus=brewer.pal(4, "Spectral")),
         main="Metagene Components - All Contributing Genes", Rowv=TRUE)
s = featureScore(n)
summary(s)
s = extractFeatures(n)
str(s)
z = featureScore(n)


numfeatures <- sapply(seq(0,1,0.01), function(num) {
  lapply(extractFeatures(n, num), function(x){
    length(x)
  })
})
minnumfeatures=20
rel.basis.contrib.cutoff <- max(seq(0,1,0.01)[apply(numfeatures,2, function(x) all(x>minnumfeatures))])

basismap(n, scale="r1", subsetRow=rel.basis.contrib.cutoff,annColors=list(basis=cbPalette[4:8], consensus=brewer.pal(4, "Spectral")), main="Metagene Components - Most Specific Genes")

coefmap(n, scale="c1", labCol=tissue_metadata$tissue, annColors=list(basis=cbPalette[4:8]))

# write out the metagene features

fs = featureScore(n)
features = extractFeatures(n, 0.90)
head_features = fs[features[[5]]]
head_features[is.na(head_features)] = 1
head_counts = tissue_only[names(head_features), colnames(tissue_only) %in% subset(summarydata, tissue == "head")$Name]
head_scores = head_features * log(rowMeans(head_counts))
head_plot = sort(head_scores, decreasing=TRUE)[1:50]
melted = melt(tissue_only[names(head_plot),])
m = merge(melted, summarydata, by.x="X2", by.y="Name")
ggplot(m, aes(X1, value, color=tissue)) + geom_point() + scale_y_log10() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("") + ylab("score")
write.table(head_plot, file="head_metagene_scores.txt", quote=FALSE, col.names=FALSE)


atrium_features = fs[features[[1]]]
atrium_features[is.na(atrium_features)] = 1
atrium_counts = tissue_only[names(atrium_features), colnames(tissue_only) %in% subset(summarydata, tissue == "atrium")$Name]
atrium_scores = atrium_features * log(rowMeans(atrium_counts))
atrium_plot = sort(atrium_scores, decreasing=TRUE)[1:50]
melted = melt(tissue_only[names(atrium_plot),])
m = merge(melted, summarydata, by.x="X2", by.y="Name")
ggplot(m, aes(X1, value, color=tissue)) + geom_point() + scale_y_log10() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("") + ylab("score")
write.table(atrium_plot, file="atrium_metagene_scores.txt", quote=FALSE, col.names=FALSE)


spermatheca_features = fs[features[[4]]]
spermatheca_features[is.na(spermatheca_features)] = 1
spermatheca_counts = tissue_only[names(spermatheca_features), colnames(tissue_only) %in% subset(summarydata, tissue == "spermatheca")$Name]
spermatheca_scores = spermatheca_features * log(rowMeans(spermatheca_counts))
spermatheca_plot = sort(spermatheca_scores, decreasing=TRUE)[1:50]
melted = melt(tissue_only[names(spermatheca_plot),])
m = merge(melted, summarydata, by.x="X2", by.y="Name")
ggplot(m, aes(X1, value, color=tissue)) + geom_point() + scale_y_log10() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("") + ylab("score")
write.table(spermatheca_plot, file="spermatheca_metagene_scores.txt", quote=FALSE, col.names=FALSE)

MAG_features = fs[features[[3]]]
MAG_features[is.na(MAG_features)] = 1
MAG_counts = tissue_only[names(MAG_features), colnames(tissue_only) %in% subset(summarydata, tissue == "MAG")$Name]
MAG_scores = MAG_features * log(rowMeans(MAG_counts))
MAG_plot = sort(MAG_scores, decreasing=TRUE)[1:50]
melted = melt(tissue_only[names(MAG_plot),])
m = merge(melted, summarydata, by.x="X2", by.y="Name")
ggplot(m, aes(X1, value, color=tissue)) + geom_point() + scale_y_log10() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("") + ylab("score")
write.table(MAG_plot, file="MAG_metagene_scores.txt", quote=FALSE, col.names=FALSE)

body_features = fs[features[[2]]]
body_features[is.na(body_features)] = 1
body_counts = tissue_only[names(body_features), colnames(tissue_only) %in% subset(summarydata, tissue == "body")$Name]
body_scores = body_features * log(rowMeans(body_counts))
body_plot = sort(body_scores, decreasing=TRUE)[1:50]
melted = melt(tissue_only[names(body_plot),])
m = merge(melted, summarydata, by.x="X2", by.y="Name")
ggplot(m, aes(X1, value, color=tissue)) + geom_point() + scale_y_log10() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("") + ylab("score")
write.table(body_plot, file="body_metagene_scores.txt", quote=FALSE, col.names=FALSE)
```

We can see that the metagenes separate out the different body parts for the most
part, except for a single body sample which has a signature more similar to
the atrium than the body samples. Could that sample have both atrium and the rest of
body in it?

# Digging into AALB000422, AALB000423 and AALB014036
Why are 422 and 423 missing? It looks like they are only expressed in MAG, and are expressed at an extremely high level in MAG.

```{r missing-genes}
missing_genes = c("AALB000422", "AALB000423")
counts[missing_genes,]
MAG[missing_genes,]
```

We can see there is a high SE for the log fold change, which is why these are not being
called significant, it isn't that these genes aren't expressed.

```{r old-results}
old_counts = read.table("/Volumes/Clotho/Users/rory/cache/mosquito_project/data/albimanus/combined.counts", header=TRUE, sep="\t", row.names=1)
colnames(old_counts) = c("AT3_MAG_M_Albi", "BT3_ALBI_At_M", "BT3_ALBI_At_V",
 "BT3_MAG_V_Albi", "CT3_ALBI_At_M",  "CT3_ALBI_At_V",  "CT3_MAG_M_Albi", "CT3_MAG_V_Albi",
 "DT3_ALBI_At_M",  "DT3_ALBI_At_V",  "DT3_MAG_M_Albi", "DT3_MAG_V_Albi")
old_MAG = read.table("/Volumes/Clotho/Users/rory/cache/mosquito_project/scripts/albimanus_MAG.tsv",
                     sep="\t", header=TRUE)
library(reshape)
melted_counts = melt(as.matrix(counts))
colnames(melted_counts) = c("gene", "sample", "count")
melted_counts$run = "new"

old_melted_counts = melt(as.matrix(old_counts))
colnames(old_melted_counts) = c("gene", "sample", "count")
old_melted_counts$run = "old"
head(old_melted_counts)

in_both = rbind(melted_counts[melted_counts$sample %in% old_melted_counts$sample,], old_melted_counts)
mcompare = merge(melted_counts, old_melted_counts, by.x=c("gene", "sample"), by.y=c("gene", "sample"),
 all=FALSE)
```

We can see below that in general we have more reads aligning to each gene with the new
analysis. This is likely to be partially due improvements in the aligner and improvements
in the gene models.

```{r old-vs-new}
ggplot(mcompare, aes(count.x, count.y)) + geom_point() + xlab("new") + ylab("old") + scale_y_sqrt() + scale_x_sqrt()
```

Looking at the old code that just had the atrium and MAG to compare, we did the same thing where we built the model using all of the samples. We used limma last time to call differential expression instead of DESeq2 which is another source of difference; we could do the same thing again for this sample and see if it is a DESeq2 vs limma difference or not.

```{r mag-limma}
library(edgeR)
design = model.matrix(~ 0 + tissue_status, data=summarydata)
colnames(design) = gsub("tissue_status", "", colnames(design))
y = DGEList(counts=counts)
y = calcNormFactors(y)
v = voom(y, design, plot=TRUE)
fit_limma = lmFit(v, design)
cm = makeContrasts(atrium = atrium_mated - atrium_virgin,
                   MAG = MAG_mated - MAG_virgin,
                   head = head_mated - head_virgin,
                   spermatheca = spermatheca_mated - spermatheca_virgin, levels=design)
fit_limma2 = contrasts.fit(fit_limma, cm)
fit_limma2 = eBayes(fit_limma2)
MAG_results_limma = topTable(fit_limma2, n=Inf, p.value=1, coef="MAG")
MAG_results_limma[missing_genes,]
```

ALB000422 gets knocked out of being significant if you use limma instead of
DESeq2, but AALB000423 doesn't. If we just do what we did before, include
only the atirum and the MAG results, they are both significant.

```{r mag-limma-atrium-MAG-only}
library(edgeR)
summarysubset = subset(summarydata, tissue %in% c("atrium", "MAG"))
countssubset = counts[, colnames(counts) %in% summarysubset$Name]
summarysubset$tissue_status = as.character(summarysubset$tissue_status)
summarysubset$tissue = as.character(summarysubset$tissue)
summarysubset$status = as.character(summarysubset$status)
design = model.matrix(~ 0 + tissue_status, data=summarysubset)
colnames(design) = gsub("tissue_status", "", colnames(design))
y = DGEList(counts=countssubset)
y = calcNormFactors(y)
v = voom(y, design, plot=FALSE)
fit_limma_subset = lmFit(v, design)
cm = makeContrasts(atrium = atrium_mated - atrium_virgin,
                   MAG = MAG_mated - MAG_virgin, levels=design)
fit_limma_subset2 = contrasts.fit(fit_limma_subset, cm)
fit_limma_subset2 = eBayes(fit_limma_subset2)
MAG_results_limma_subset = topTable(fit_limma_subset2, n=Inf, p.value=1, coef="MAG")
MAG_results_limma_subset[missing_genes,]
```

However quantile normalizing the voom counts also has an effect:

```{r mag-limma-atrium-normalized}
library(edgeR)
summarysubset = subset(summarydata, tissue %in% c("atrium", "MAG"))
countssubset = counts[, colnames(counts) %in% summarysubset$Name]
summarysubset$tissue_status = as.character(summarysubset$tissue_status)
summarysubset$tissue = as.character(summarysubset$tissue)
summarysubset$status = as.character(summarysubset$status)
design = model.matrix(~ 0 + tissue_status, data=summarysubset)
colnames(design) = gsub("tissue_status", "", colnames(design))
y = DGEList(counts=countssubset)
y = calcNormFactors(y)
v = voom(y, design, plot=FALSE, normalize='quantile')
fit_limma_subset_quantile = lmFit(v, design)
cm = makeContrasts(atrium = atrium_mated - atrium_virgin,
                   MAG = MAG_mated - MAG_virgin, levels=design)
fit_limma_subset_quantile2 = contrasts.fit(fit_limma_subset_quantile, cm)
fit_limma_subset_quantile2 = eBayes(fit_limma_subset_quantile2)
MAG_results_limma_subset_quantile = topTable(fit_limma_subset_quantile2, n=Inf, p.value=1, coef="MAG")
MAG_results_limma_subset_quantile[missing_genes,]
```

So there are a set of genes that are pretty sensitive to how we do the analysis;
if we swap callers, or how we combine them together or how we do the normalization,
it matters. These are a set of genes that are not robust changes. We can make
reasonable arguments for doing any of these.


```{r reproducible-hits}
MAG <- MAG[order(rownames(MAG)),]
MAG_results_limma <- MAG_results_limma[order(rownames(MAG_results_limma)),]
MAG_results_limma_subset <- MAG_results_limma_subset[order(rownames(MAG_results_limma_subset)),]
MAG_results_limma_subset_quantile <- MAG_results_limma_subset_quantile[order(rownames(MAG_results_limma_subset_quantile)),]
```

The actual fold changes between doing an analysis on the reduced dataset vs.
the full one are not much different.

```{r fold-change-subset}
qplot(MAG_results_limma$logFC, MAG_results_limma_subset$logFC) +
   xlab("full model") + ylab("reduced model") +
  theme_bw() +
  theme(text=element_text(family="Gill Sans", size=10),
        strip.background = element_rect(fill="white"))

qplot((MAG_results_limma$logFC + MAG_results_limma_subset$logFC)/2,
  MAG_results_limma$logFC - MAG_results_limma_subset$logFC, geom="hex") + xlab("average logFC") +
  ylab("full logFC - reduced logFC") +
  theme_bw() +
  theme(text=element_text(family="Gill Sans", size=10),
        strip.background = element_rect(fill="white")) +
  ggtitle("bland-altman plot")
```

But the calculated FDR is very different:

```{r padj-subset}
qplot(MAG_results_limma$adj.P.Val, MAG_results_limma_subset$adj.P.Val) +
    xlab("full model") + ylab("reduced model")
```

```{r correlation-plots}
qplot((MAG_results_limma$logFC + MAG_results_limma_subset$logFC)/2,
  MAG_results_limma$logFC - MAG_results_limma_subset$logFC) + xlab("average logFC") +
    ylab("full logFC - reduced logFC")
qplot((MAG_results_limma$P.Value + MAG_results_limma_subset$P.Value)/2,
  MAG_results_limma$P.Value - MAG_results_limma_subset$P.Value) + xlab("average pvalue") + ylab("full pvalue - reduced pvalue")

qplot((fit_limma$sigma+ fit_limma_subset$sigma)/2, fit_limma$sigma-fit_limma_subset$sigma) + stat_binhex() + scale_x_sqrt() +
  xlab("average variance") + ylab("full - reduced variance")

in_either = unique(c(rownames(subset(MAG_results_limma, adj.P.Val < 0.1)),
    rownames(subset(MAG_results_limma_subset, adj.P.Val < 0.1))))

signew = MAG_results_limma[in_either,]
signew$run = "full"
signew$sig = signew$adj.P.Val < 0.1
sigold = MAG_results_limma_subset[in_either,]
sigold$run = "subset"
sigold$sig = sigold$adj.P.Val < 0.1

m = merge(signew, sigold, by="row.names")
m$both = ifelse(m$sig.x & m$sig.y, "both", ifelse(m$sig.x, "full only", "subset only"))
m$minpval <- pmin(m$adj.P.Val.x, m$adj.P.Val.y)
ggplot(m, aes((AveExpr.x + AveExpr.y)/2, (AveExpr.x - AveExpr.y),
              size=minpval)) + geom_point(alpha=0.6) + xlab("average expression") + ylab("expression difference") + facet_wrap(~ both) +
scale_size(range=c(0.3,3)) + element_blank() +
guides(colour = guide_legend(override.aes = list(alpha = 1),
                             title="found in"))

allsig = rbind(signew, sigold)
allsig$sig = allsig$adj.P.Val < 0.1
allsig$id = rownames(allsig)
library(dplyr)

ggplot(allsig, aes(AveExpr, logFC, shape=run, color=sig)) + geom_point()
```

# limma
Write out the counts and subsetted summardata file for use with limma.

```{r limma-write}
write.table(file="limma.counts", counts, col.names=TRUE, row.names=TRUE,
quote=FALSE, sep="\t")
write.table(file="limma.summarydata.txt", summarydata, col.names=TRUE,
row.names=TRUE, quote=FALSE, sep="\t")
```
